### PR TITLE
Cloud provider cleanup

### LINKED ISSUE #1933: AWS create cluster fail: 'dict object' has no attribute 'ec2_tag_sub-host-type'
Hi.

When running a fresh installation with 

bin/cluster create aws dev

It stops at:

TASK: [openshift_facts ] ****************************************************** 
fatal: [dev-node-compute-e97ed] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [dev-node-compute-38774] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [dev-node-infra-d4f02] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'

I have checked and instances has sub-host-type tag.

Any clue?

### PR DESCRIPTION
**Pull request by detiber**
Fixes up the way we are using some variables like ec2_region

Fixes https://github.com/openshift/openshift-ansible/issues/1760
Fixes https://github.com/openshift/openshift-ansible/issues/1826
Fixes https://github.com/openshift/openshift-ansible/issues/1825
Fixes #1933

Also fixes issues with using vars_files on an include.

### CODE REVIEW

**Comment by detiber** [comment]
@etsauer I'm testing this now.
@twiest heads up on these changes.
@lhuard1A wanted to get your opinion on these changes as well

**Comment by twiest** [comment]
@detiber thanks for the heads up. These all look fine to me.

@mwoodson heads up :)

**Comment by etsauer** [comment]
@detiber still getting the issue below with the ec2_region var on the nodes:

```
TASK: [openshift_facts ] ****************************************************** 
fatal: [esauer-test6-node-infra-814d2] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [esauer-test6-node-compute-9c57d] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [esauer-test6-node-compute-43f88] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/root/launch.retry

esauer-test6-master-47bdf  : ok=246  changed=46   unreachable=0    failed=0   
esauer-test6-node-compute-43f88 : ok=38   changed=7    unreachable=1    failed=0   
esauer-test6-node-compute-9c57d : ok=38   changed=7    unreachable=1    failed=0   
esauer-test6-node-infra-814d2 : ok=38   changed=7    unreachable=1    failed=0   
localhost                  : ok=85   changed=9    unreachable=0    failed=0   

ACTION [create] failed: Command 'ansible-playbook  -i inventory/aws/hosts -e 'num_masters=1 num_nodes=2 cluster_id=esauer-test6 cluster_env=dev num_etcd=0 num_infra=1 deployment_type=openshift-enterprise' playbooks/aws/openshift-cluster/launch.yml' returned non-zero exit status 3
```

**Comment by detiber** [comment]
@etsauer you are getting further than me at this point. I'm still trying to get my etcd hosts to provision correctly (for some reason they are not getting the requiretty change properly)

**Comment by etsauer** [comment]
```
TASK: [openshift_facts ] ****************************************************** 
fatal: [esauer-test6-node-infra-814d2] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [esauer-test6-node-compute-9c57d] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'
fatal: [esauer-test6-node-compute-43f88] => One or more undefined variables: 'dict object' has no attribute 'ec2_tag_sub-host-type'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/root/launch.retry

esauer-test6-master-47bdf  : ok=246  changed=46   unreachable=0    failed=0   
esauer-test6-node-compute-43f88 : ok=38   changed=7    unreachable=1    failed=0   
esauer-test6-node-compute-9c57d : ok=38   changed=7    unreachable=1    failed=0   
esauer-test6-node-infra-814d2 : ok=38   changed=7    unreachable=1    failed=0   
localhost                  : ok=85   changed=9    unreachable=0    failed=0   

ACTION [create] failed: Command 'ansible-playbook  -i inventory/aws/hosts -e 'num_masters=1 num_nodes=2 cluster_id=esauer-test6 cluster_env=dev num_etcd=0 num_infra=1 deployment_type=openshift-enterprise' playbooks/aws/openshift-cluster/launch.yml' returned non-zero exit status 3
```

**Comment by etsauer** [comment]
@detiber realized this is a new variable... what is `ec2_tag_sub-host-type`?

**Comment by etsauer** [comment]
@detiber i have not yet tried with separate etcd hosts.

**Comment by lhuard1A** [comment]
@detiber : I cannot test on AWS, but those changes look fine for me.

**Comment by detiber** [comment]
@lhuard1A which playbooks are currently in use by your team? We will be re-organizing officially supported vs. community contributed/supported content within the next few weeks.

**Comment by metmajer** [comment]
@detiber: facing the same issue as @etsauer when testing [detiber:aws_cleanup](https://github.com/detiber/openshift-ansible/tree/aws_cleanup).

**Comment by metmajer** [comment]
I managed to temporarily overcome this issue by setting `openshift_node_labels.type: unknown` in `playbooks/aws/openshift-cluster/config.yml`. Ran smoothly for a while, but then I got the following issue:

```
TASK: [openshift_node_dnsmasq | Install dnsmasq configuration] ****************
changed: [dev-node-compute-052d7]
changed: [dev-node-compute-aa750]
changed: [dev-node-infra-36c91]

TASK: [openshift_node_dnsmasq | Install network manager dispatch script] ******
skipping: [dev-node-infra-36c91]
skipping: [dev-node-compute-052d7]
skipping: [dev-node-compute-aa750]

TASK: [openshift_node_dnsmasq | fail msg="Not implemented"] *******************
failed: [dev-node-compute-052d7] => {"failed": true}
msg: Not implemented
failed: [dev-node-compute-aa750] => {"failed": true}
msg: Not implemented
failed: [dev-node-infra-36c91] => {"failed": true}
msg: Not implemented

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/cwat-metmajer/launch.retry

dev-master-d8fbd           : ok=264  changed=48   unreachable=0    failed=0
dev-node-compute-052d7     : ok=93   changed=21   unreachable=0    failed=1
dev-node-compute-aa750     : ok=93   changed=21   unreachable=0    failed=1
dev-node-infra-36c91       : ok=93   changed=21   unreachable=0    failed=1
localhost                  : ok=88   changed=9    unreachable=0    failed=0
```

**Comment by metmajer** [comment]
I successfully create a cluster by adding `openshift_use_dnsmasq: no` to `playbooks/aws/openshift-cluster/vars.yml`. Fow now, overriding this variable has been suggested in #1807.

**Comment by detiber** [comment]
@etsauer @metmajer I'm currently testing the latest changes, but I believe the issues you've hit should now be addressed.

**Comment by detiber** [comment]
Nevermind, still have the oo_nodes_to_config error to address...

**Comment by detiber** [comment]
@etsauer @metmajer This branch is now passing my tests. Let me know if there are hitting any additional issues.

**Comment by metmajer** [comment]
I just tried a test run, but failed:

```
export ec2_image=ami-6d1c2007
export ec2_keypair=openshift-origin-dev
export ec2_security_groups=openshift-dev
export ec2_vpc_subnet=subnet-f1234abc
./bin/cluster create -t origin aws dev
```

resulted in:

```
TASK: [set_fact ] *************************************************************
ok: [dev-master-65321]

TASK: [set_fact ] *************************************************************
skipping: [dev-master-65321]

PLAY [Configure master certificates] ******************************************

GATHERING FACTS ***************************************************************
FATAL: no hosts matched or all hosts have already failed -- aborting


TASK: [openshift_facts | Verify Ansible version is greater than or equal to 1.9.4 and less than 2.0] ***
FATAL: no hosts matched or all hosts have already failed -- aborting
```

**Comment by detiber** [comment]
@metmajer Could you provide the full ansible output? I suspect there was some type of a failure earlier in the run that is causing the "no hosts matched or all hosts have already failed -- aborting" message in the snippet you provided.

**Comment by metmajer** [comment]
@detiber here's the entire ansible run with the requested failure highlighted: https://gist.github.com/metmajer/f14a08fbfa65df1b5b468de560b6dba6#file-openshift-ansible-run-L1241-L1243.

**Comment by detiber** [comment]
@metmajer ah, I think I see the issue here. It looks like you are picking up hosts from a previous deployment.

If you encounter a failed run using bin/cluster you have a couple of options to proceed:
- you can re-run using the config option
```
./bin/cluster config -t origin aws dev
```
- you can terminate the old cluster and recreate
```
./bin/cluster terminate -t origin aws dev
# refreshing the inventory cache will prevent the terminated hosts from being picked up on a subsequent run
inventory/aws/hosts/ec2.py --refresh-cache
./bin/cluster create -t origin aws dev
```

If an error happens before all of the hosts have been launched, then you will need to use the terminate/create options, since the config option will not spawn new hosts.

As an aside, you should not need to pass `-t origin`, since that should be the default deployment_type if none is specified.

**Comment by metmajer** [comment]
Good catch @detiber! What you proposed did the trick for a while, but then another failure popped up:

```
TASK: [openshift_facts ] ******************************************************
fatal: [dev-node-compute-68df4] => One or more undefined variables: 'ec2_region' is undefined
fatal: [dev-node-compute-54c64] => One or more undefined variables: 'ec2_region' is undefined
fatal: [dev-node-infra-d4ba2] => One or more undefined variables: 'ec2_region' is undefined
```

See the full ansible run here with failures highlighted here: https://gist.github.com/metmajer/f14a08fbfa65df1b5b468de560b6dba6#file-openshift-ansible-run2-L2732-L2734. Thanks!

**Comment by detiber** [comment]
@metmajer I'm not able to replicate the last error the "'ec2_region' is undefined" error using this branch. I'm even having trouble figuring out how the code in this branch could cause that error now. It looks like the only reference to `ec2_region` in the `create` workflow is through the oo_option lookup, which wouldn't trigger that error.

**Comment by metmajer** [comment]
Apologies, @detiber. I accidentially ended up in the `master` branch of your fork. The cluster provisioning went smoothly. Thanks for the quick fix.

Any ideas whether your branch could affect https://github.com/openshift/openshift-ansible/issues/1837?

**Comment by detiber** [comment]
ok to test

**Comment by jannispl** [comment]
Could we also get rid of this hardcoded flavor which is not necessarily available on custom OpenStack deployments?
https://github.com/openshift/openshift-ansible/blob/master/playbooks/openstack/openshift-cluster/launch.yml#L49

File: playbooks/openstack/openshift-cluster/files/user-data [hunk 1]
```diff
@@ -1,13 +0,0 @@
...
-system_info:
-  default_user:
-    name: openshift
-    sudo: ["ALL=(ALL) NOPASSWD: ALL"]
-
-write_files:
-  - path: /etc/sudoers.d/00-openshift-no-requiretty
-    permissions: 440
-    content: |
-      Defaults:openshift !requiretty
```
Comment (jannispl) [inline playbooks/openstack/openshift-cluster/files/user-data]: `user-data` file is still being referenced from heat_stack_server.yml

Comment (detiber) [inline playbooks/openstack/openshift-cluster/files/user-data]: @jannispl Thanks! I don't currently have an OpenStack env to test against, so I fully expected there to be something off.

**Comment by detiber** [comment]
@jannispl I believe I've taken care of the hardcoded dns_flavor now as well. Please let me know if there an any other issues with the openstack deployments.

**Comment by jannispl** [comment]
@detiber I've been using your branch to deploy to OpenStack since the current master is non-functional in that department. The only issues I was still having: the internal subnet is created with the DNS Server pointing to the internal dns instance, however when ansible tries to install bind here, something can not be resolved (and there's no other dns available) so I have to quickly add 8.8.8.8 as second nameserver to the subnet in the openstack dashboard after the template is created and before ansible reaches dns server setup. 
The other problem was openshift_use_dnsmasq needed to be false in playbooks/openstack/../config.yaml, which I've seen you have done in the aws section but not openstack. Might want to add this to openstack as well. Other than that, the deployment went smooth. (I deployed single-master with 3 normal nodes) If you need anything else to be tested on openstack, just ping me.

**Pull request merged**